python/ray/autoscaler/resource_demand_scheduler.py

```diff
@@ -0,0 +1,197 @@
+import copy
+import numpy as np
+import logging
+import collections
+from typing import List, Dict, Tuple
+
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import TAG_RAY_INSTANCE_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class ResourceDemandScheduler:
+    def __init__(self, provider: NodeProvider, instance_types: Dict[str, dict],
+                 max_workers: int):
+        self.provider = provider
+        self.instance_types = instance_types
+        self.max_workers = max_workers
+
+    def debug_string(self, nodes: List[str],
+                     pending_nodes: Dict[str, int]) -> str:
+        node_resources, instance_type_counts = self.calculate_node_resources(
+            nodes, pending_nodes)
+
+        out = "Worker instance types:"
+        for instance_type, count in instance_type_counts.items():
+            out += "\n - {}: {}".format(instance_type, count)
+            if pending_nodes.get(instance_type):
+                out += " ({} pending)".format(pending_nodes[instance_type])
+
+        return out
+
+    def calculate_node_resources(
+            self, nodes: List[str],
+            pending_nodes: Dict[str, int]) -> (List[dict], Dict[str, int]):
+        """Returns node resource list and instance type counts."""
+
+        node_resources = []
+        instance_type_counts = collections.defaultdict(int)
+
+        def add_instance(instance_type):
+            if instance_type not in self.instance_types:
+                raise RuntimeError(
+                    "Missing entry for instance_type {} in "
+                    "available_instance_types config: {}".format(
+                        instance_type, self.instance_types))
+            # Careful not to include the same dict object multiple times.
+            node_resources.append(
+                copy.deepcopy(self.instance_types[instance_type]["resources"]))
+            instance_type_counts[instance_type] += 1
+
+        for node_id in nodes:
+            tags = self.provider.node_tags(node_id)
+            if TAG_RAY_INSTANCE_TYPE in tags:
+                instance_type = tags[TAG_RAY_INSTANCE_TYPE]
+                add_instance(instance_type)
+
+        for instance_type, count in pending_nodes.items():
+            for _ in range(count):
+                add_instance(instance_type)
+
+        return node_resources, instance_type_counts
+
+    def get_instances_to_launch(self, nodes: List[str],
```

Comment:
I think this is fine since these logs aren't user visible and only run a few times a minute.. It probably at most doubles the number of log message for the autoscaler.